pkg/agent/proxy/proxier_test.go

```diff
@@ -752,193 +906,252 @@ func testClusterIPRemoval(t *testing.T, svcIP net.IP, epIP net.IP, isIPv6 bool)
 	fp.syncProxyRules()
 
-	fp.endpointsChanges.OnEndpointUpdate(epUDP, nil)
+	fp.serviceChanges.OnServiceUpdate(svc, nil)
+	fp.endpointsChanges.OnEndpointUpdate(ep, nil)
 	fp.syncProxyRules()
 }
 
-func TestClusterIPRemoveSamePortEndpointIPv4(t *testing.T) {
-	testClusterIPRemoveSamePortEndpoint(t, net.ParseIP("10.20.30.41"), net.ParseIP("10.180.0.1"), false)
+func TestClusterIPRemove(t *testing.T) {
+	t.Run("IPv4", func(t *testing.T) {
+		testClusterIPRemove(t, svc1IPv4, ep1IPv4, false)
+	})
+	t.Run("IPv6", func(t *testing.T) {
+		testClusterIPRemove(t, svc1IPv6, ep1IPv6, true)
+	})
 }
 
-func TestClusterIPRemoveSamePortEndpointIPv6(t *testing.T) {
-	testClusterIPRemoveSamePortEndpoint(t, net.ParseIP("10:20::41"), net.ParseIP("10:180::1"), true)
+func TestNodePortRemove(t *testing.T) {
+	t.Run("IPv4", func(t *testing.T) {
+		testNodePortRemove(t, nodePortAddressesIPv4, svc1IPv4, ep1IPv4, false)
+	})
+	t.Run("IPv6", func(t *testing.T) {
+		testNodePortRemove(t, nodePortAddressesIPv6, svc1IPv6, ep1IPv6, true)
+	})
 }
 
-func testClusterIPRemoveEndpoints(t *testing.T, svcIP net.IP, epIP net.IP, isIPv6 bool) {
+func TestLoadBalancerRemove(t *testing.T) {
+	t.Run("IPv4", func(t *testing.T) {
+		testLoadBalancerRemove(t, nodePortAddressesIPv4, svc1IPv4, ep1IPv4, loadBalancerIPv4, false)
+	})
+	t.Run("IPv6", func(t *testing.T) {
+		testLoadBalancerRemove(t, nodePortAddressesIPv6, svc1IPv6, ep1IPv6, loadBalancerIPv6, true)
+	})
+
+}
+
+func testClusterIPNoEndpoint(t *testing.T, svcIP net.IP, isIPv6 bool) {
 	ctrl := gomock.NewController(t)
 	defer ctrl.Finish()
 	mockOFClient := ofmock.NewMockClient(ctrl)
 	mockRouteClient := routemock.NewMockInterface(ctrl)
 	groupAllocator := openflow.NewGroupAllocator(isIPv6)
 	fp := NewFakeProxier(mockRouteClient, mockOFClient, nil, groupAllocator, isIPv6)
 
-	svcPort := 80
-	svcPortName := k8sproxy.ServicePortName{
-		NamespacedName: makeNamespaceName("ns1", "svc1"),
-		Port:           "80",
-		Protocol:       corev1.ProtocolTCP,
-	}
-	makeServiceMap(fp,
-		makeTestService(svcPortName.Namespace, svcPortName.Name, func(svc *corev1.Service) {
-			svc.Spec.ClusterIP = svcIP.String()
-			svc.Spec.Ports = []corev1.ServicePort{{
-				Name:     svcPortName.Port,
-				Port:     int32(svcPort),
-				Protocol: corev1.ProtocolTCP,
-			}}
-		}),
-	)
-
-	ep := makeTestEndpoints(svcPortName.Namespace, svcPortName.Name, func(ept *corev1.Endpoints) {
-		ept.Subsets = []corev1.EndpointSubset{{
-			Addresses: []corev1.EndpointAddress{{
-				IP: epIP.String(),
-			}},
-			Ports: []corev1.EndpointPort{{
-				Name:     svcPortName.Port,
-				Port:     int32(svcPort),
-				Protocol: corev1.ProtocolTCP,
-			}},
-		}}
+	svcPortName := makeSvcPortName("ns", "svc", strconv.Itoa(svcPort), corev1.ProtocolTCP)
+	svc := makeTestClusterIPService(&svcPortName, svcIP, int32(svcPort), corev1.ProtocolTCP, nil, nil)
+	makeServiceMap(fp, svc)
+	makeEndpointsMap(fp)
+	fp.syncProxyRules()
```

Comment:
What's the expectation of this test? If it's no-ops, I think it should be covered by other cases. Could you check?

Reply:
Without Endpoint, just no expectation.